ext/php8/handlers_curl.c

```diff
@@ -79,54 +76,44 @@ static void dd_ch_duplicate_headers(zval *ch_orig, zval *ch_new) {
     }
 }
 
-static void dd_init_headers_arg(zval *arg, zval *ch) {
-    HashTable *retval;
-    ALLOC_HASHTABLE(retval);
-    HashTable *headers = NULL;
-
-    if (dd_headers) {
-        headers = zend_hash_index_find_ptr(dd_headers, Z_OBJ_HANDLE_P(ch));
-        if (headers) {
-            size_t headers_count = zend_hash_num_elements(headers);
-            zend_hash_init(retval, headers_count, NULL, ZVAL_PTR_DTOR, 0);
-            zend_hash_copy(retval, headers, (copy_ctor_func_t)zval_add_ref);
-        }
+static void dd_inject_distributed_tracing_headers(zval *ch) {
+    zval headers;
+    zend_array *dd_header_array;
+    if (dd_headers && (dd_header_array = zend_hash_index_find_ptr(dd_headers, Z_OBJ_HANDLE_P(ch)))) {
+        ZVAL_ARR(&headers, zend_array_dup(dd_header_array));
+    } else {
+        array_init(&headers);
     }
 
-    if (!headers) {
-        zend_hash_init(retval, 0, NULL, ZVAL_PTR_DTOR, 0);
+    int sampling_priority;
+    if (ddtrace_fetch_prioritySampling_from_root(&sampling_priority)) {
+        add_next_index_str(&headers, zend_strpprintf(0, "x-datadog-sampling-priority: %d", sampling_priority));
+    }
+    if (DDTRACE_G(trace_id)) {
+        add_next_index_str(&headers, zend_strpprintf(0, "x-datadog-trace-id: %" PRIu64, (DDTRACE_G(trace_id))));
+    }
+    if (DDTRACE_G(span_ids_top)) {
+        add_next_index_str(&headers,
+                           zend_strpprintf(0, "x-datadog-parent-id: %" PRIu64, (DDTRACE_G(span_ids_top)->id)));
+    }
+    if (DDTRACE_G(dd_origin)) {
+        add_next_index_str(&headers, zend_strpprintf(0, "x-datadog-origin: %s", ZSTR_VAL(DDTRACE_G(dd_origin))));
     }
-    ZVAL_ARR(arg, retval);
-}
 
-static void dd_free_headers_arg(zval *arg) { zend_array_destroy(Z_ARRVAL_P(arg)); }
+    zend_function *setopt_fn = zend_hash_str_find_ptr(EG(function_table), ZEND_STRL("curl_setopt"));
+    zend_constant *http_header_const = zend_hash_str_find_ptr(EG(zend_constants), ZEND_STRL("CURLOPT_HTTPHEADER"));
```

Comment:
Any particular reason this needs to be fetched again for the call, or can we use the value of `dd_const_curlopt_httpheader` that is fetched on startup?

Reply:
Wasn't aware this exists, will change :-)